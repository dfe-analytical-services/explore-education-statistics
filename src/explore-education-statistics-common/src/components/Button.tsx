import classNames from 'classnames';
import React, { MouseEventHandler, ReactNode } from 'react';
import styles from './Button.module.scss';

export interface ButtonProps {
  children: ReactNode;
  className?: string;
  disabled?: boolean;
  id?: string;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  variant?: 'secondary' | 'warning';
  type?: 'button' | 'submit' | 'reset';
}

const Button = ({
  children,
  className,
  id,
  onClick,
  disabled = false,
  variant,
  type = 'button',
}: ButtonProps) => {
  return (
    // eslint-disable-next-line react/button-has-type
    <button
      aria-disabled={disabled}
      className={classNames(
        'govuk-button',
        styles.button,
        {
          'govuk-button--disabled': disabled,
          'govuk-button--secondary': variant === 'secondary',
          'govuk-button--warning': variant === 'warning',
        },
        className,
      )}
      disabled={disabled}
      id={id}
      onClick={onClick}
      type={type}
    >
      {children}
    </button>
  );
};

export default Button;
