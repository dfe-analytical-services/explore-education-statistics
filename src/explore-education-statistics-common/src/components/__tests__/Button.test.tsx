import flushPromises from '@common-test/flushPromises';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import React from 'react';
import Button from '../Button';

describe('Button', () => {
  test('renders correctly with required props', () => {
    const { container } = render(<Button>Test button</Button>);

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('correctly sets attributes when `disabled = true`', () => {
    render(<Button disabled>Test button</Button>);

    const button = screen.getByRole('button', { name: 'Test button' });

    expect(button).toBeDisabled();
    expect(button).toBeAriaDisabled();
    expect(button).not.toHaveAttribute('aria-describedby');
  });

  test('correctly sets attributes when `ariaDisabled = true`', () => {
    render(
      <Button ariaDisabled id="test-button">
        Test button
      </Button>,
    );

    const button = screen.getByRole('button', { name: 'Test button' });

    expect(button).not.toBeDisabled();
    expect(button).toBeAriaDisabled();
  });

  test('correctly sets attributes when using both `disabled` and `ariaDisabled`', () => {
    render(
      <Button disabled ariaDisabled id="test-button">
        Test button
      </Button>,
    );

    const button = screen.getByRole('button', { name: 'Test button' });

    expect(button).not.toBeDisabled();
    expect(button).toBeAriaDisabled();
  });

  test('disabled if current `onClick` handler is processing', async () => {
    jest.useFakeTimers({
      legacyFakeTimers: true,
    });

    const handleClick = jest.fn(
      () => new Promise(resolve => setTimeout(resolve, 2000)),
    );

    render(<Button onClick={handleClick}>Test button</Button>);

    const button = screen.getByRole('button', { name: 'Test button' });

    expect(button).toBeEnabled();

    userEvent.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
    expect(button).toBeDisabled();

    jest.advanceTimersByTime(1000);

    // Button is still disabled
    expect(button).toBeDisabled();

    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  test('enabled if current `onClick` handler is processing and `disableDoubleClick` is false', async () => {
    jest.useFakeTimers();

    const handleClick = jest.fn(
      () => new Promise(resolve => setTimeout(resolve, 2000)),
    );

    render(
      <Button disableDoubleClick={false} onClick={handleClick}>
        Test button
      </Button>,
    );

    const button = screen.getByRole('button', { name: 'Test button' });

    expect(button).toBeEnabled();

    userEvent.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
    expect(button).toBeEnabled();

    jest.advanceTimersByTime(1000);
    // Flushes promise queue so any state change is triggered
    await flushPromises();

    // Button is still enabled
    expect(button).toBeEnabled();

    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  test('enabled once the current `onClick` handler has finished', async () => {
    jest.useFakeTimers();

    const handleClick = jest.fn(
      () => new Promise(resolve => setTimeout(resolve, 2000)),
    );

    render(<Button onClick={handleClick}>Test button</Button>);

    const button = screen.getByRole('button', { name: 'Test button' });

    userEvent.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
    expect(button).toBeDisabled();

    jest.advanceTimersByTime(2000);
    // Flushes promise queue so any state change is triggered
    await flushPromises();

    // Task has completed, so button is now enabled
    expect(button).toBeEnabled();

    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });
});
