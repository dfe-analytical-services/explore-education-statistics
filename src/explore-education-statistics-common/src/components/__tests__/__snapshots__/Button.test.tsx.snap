// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly with required props 1`] = `
<button aria-disabled="false"
        class="govuk-button button"
        type="button"
>
  Test button
</button>
`;
